db-tests/src/test/java/gov/nasa/jpl/aerie/database/ExternalEventTests.java

```diff
@@ -0,0 +1,2215 @@
+            // language=sql
+            """
+            DELETE FROM merlin.external_source_type WHERE name='%s';
+            """.formatted(st)
+          )
+        );
+        if (!ex.getSQLState().equals("23503")
+            && !ex.getMessage().contains(
+                "update or delete on table \"external_source_type\" violates foreign key constraint "
+                + "\"derivation_group_references_external_source_type\" on table \"derivation_group\"")
+        ) {
+          throw ex;
+        }
+      }
+    }
+
+    /**
+     * It should not be possible to delete an event type with an event still associated.
+     */
+    @Test
+    void deleteEventTypeWithRemainingEvent() throws SQLException {
+      try (final var statement = connection.createStatement()) {
+
+        // create source and event
+        ExternalSource src = new ExternalSource(
+            "A",
+            st,
+            dg,
+            "2024-01-01T00:00:00Z",
+            "2024-01-01T00:00:00Z",
+            "2024-01-01T00:30:00Z",
+            ca
+        );
+        ExternalEvent evt = createEvent("A_1", "2024-01-01T00:00:00Z", "00:05:0", src);
+
+        // insert the event and her types
+        // insert generic external event type, source type, and derivation group
+        insertStandardTypes();
+
+        // insert sources
+        insertExternalSource(src);
+
+        // insert events
+        insertExternalEvent(evt);
+
+        // delete the event type (expect error)
+        final SQLException ex = assertThrows(PSQLException.class,
+          () -> statement.executeUpdate(
+              // language=sql
+              """
+              DELETE FROM merlin.external_event_type WHERE name='%s';
+              """.formatted(et)
+          )
+        );
+        if (!ex.getSQLState().equals("23503")
+            && !ex.getMessage().contains(
+                "update or delete on table \"external_event_type\" violates foreign key constraint "
+                + "\"external_event_references_event_type_name\" on table \"external_event\"")
+        ) {
+          throw ex;
+        }
+      }
+    }
+  }
+  //endregion
+
+
+  //region Miscellaneous
+  /**
+   * Finally, we include an extra test. This is a comprehensive test for derived events that manages several derivation
+   *    groups (which entails repeating basicDerivedEvents twice but the second round bears a new DG name. Then, we
+   *    verify that there is no overlap! Note that this test is effectively vacuous but is a good sanity check.).
+   */
+  @Test
+  void superDerivedEvents() throws SQLException {
+    // upload all source data, but twice (using different dgs, to prove non overlap in derivation)
+    String dg2 = dg + "_2";
+
+    // upload the data once for the first derivation group
+    upload_source(dg);
+
+    // repeat (explicitly, for ease of implementation) with the second derivation group
```

Comment:
Judging by the comment and a brief skim, the insertions following L1929 look like they can be replace with `upload_source(dg2)`. If this is the case, this replacement is highly preferred.

Reply:
The difference was in the external source names, as well as that `upload_source` uploads the event types too. Since we are avoiding upserts in our tests, running this function twice uploads the event types twice as well, which we want to avoid. This was one of the cases when we were talking about removing upserts earlier where we decided to deal with potential repetitious code.